src/NHibernate/Cache/StandardQueryCache.cs

```diff
@@ -551,7 +551,7 @@ private static ICacheAssembler[] GuessTypes(IList cacheable)
 
 		protected virtual bool IsUpToDate(ISet<string> spaces, long timestamp)
 		{
-			return _updateTimestampsCache.IsUpToDate(spaces, timestamp);
+			return spaces.Count == 0 || _updateTimestampsCache.IsUpToDate(spaces, timestamp);
```

Comment:
It feels like better fix would be to replace `Synchronized` attributes with locking on field in `UpdateTimestampsCache` and move `spaces.Count` check inside.

Reply:
Maybe. I'll change.

Reply:
Hi @hazzik,
Also need to check spaces.count before calling Invalidate and PreInvalidate?
```c#
if (session.Factory.Settings.IsQueryCacheEnabled)
{
	session.Factory.UpdateTimestampsCache.Invalidate(executedSpaces);
}

if (session.Factory.Settings.IsQueryCacheEnabled)
{
	session.Factory.UpdateTimestampsCache.PreInvalidate(executedSpaces);
}
```

Reply:
There is a problem with hiding this logic into the method. These methods are virtual and if the logic is hidden there it brings a burden of implementation to the client. So, after thinking overnight I think the original approach is better.

I'll probably change the logic to be in both places. Inside and outside of these method calls.

Reply:
> it brings a burden of implementation to the client

IMHO it's OK to make it client burden. We shouldn't try to foresee and optimize user implementations.

Reply:
Hi @hazzik ,
I also realized that we are storing updatetimestampscache for non cache'ing entities.
Is this really necessary in our scenario we use query cache for just Cache.ReadWrite or Cache.Readonly entities?

We need high performance to do that we need to decrease Sync. and locking for common dml/select operations.

I think that this needs to be configurable(Store all entities updatetimestampscache or just cacheable entities) or we give us a way to implement our UpdateTimestampsCache?

In distributed cache scenario there is a lot of cache call(for cached entities) may be there would be a way to implement memory cache in front of second level cache?